File: frontend/src/types/stats.ts
import type { AudioQuality } from './playback';

export interface PaginationMeta {
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
  has_next: boolean;
  has_previous: boolean;
}

export interface ArtistSummary {
  id: number;
  name: string;
  sort_name?: string | null;
  created_at: string;
}

export interface AlbumSummary {
  id: number;
  title: string;
  artist_id: number;
  release_year?: number | null;
  genre?: string | null;
  created_at: string;
  cover_art_url?: string | null;
  artist?: ArtistSummary | null;
}

export interface TrackPlayStats {
  id: number;
  title: string;
  artist_id?: number | null;
  album_id?: number | null;
  track_number?: number | null;
  disc_number?: number | null;
  genre?: string | null;
  duration_seconds?: number | null;
  file_path: string;
  file_hash: string;
  created_at: string;
  updated_at: string;
  play_count: number;
  last_played_at?: string | null;
  artist?: ArtistSummary | null;
  album?: AlbumSummary | null;
}

export interface TrackPlayStatsResponse {
  items: TrackPlayStats[];
  pagination: PaginationMeta;
}

export interface AlbumListResponse {
  items: AlbumSummary[];
  pagination: PaginationMeta;
}

export interface GenreStats {
  genre: string;
  track_count: number;
  total_duration_seconds: number;
}

export interface ArtistStats {
  name: string;
  track_count: number;
  album_count: number;
}

export interface AlbumStats {
  title: string;
  artist_name?: string | null;
  release_date?: string | null;
  track_count: number;
  total_duration_seconds: number;
  total_duration_minutes: number;
}

export interface QualityStats {
  bitrate_distribution: Array<{
    bitrate: number;
    count: number;
  }>;
  sample_rate_distribution: Array<{
    sample_rate: number;
    count: number;
  }>;
}

export interface StatsOverview {
  total_albums: number;
  total_tracks: number;
  total_artists: number;
  total_duration_seconds: number;
  total_duration_hours: number;
  estimated_size_mb: number;
  average_bitrate?: number | null;
  average_track_duration?: number | null;
  recently_added_at?: string | null;
}

export interface DashboardData {
  overview: StatsOverview;
  recent_albums: AlbumSummary[];
  top_tracks: TrackPlayStats[];
  genre_distribution: GenreStats[];
  scan_status: string;
  scan_progress?: number | null;
  generated_at: string;
}

export interface MostPlayedQueryParams {
  page?: number;
  pageSize?: number;
  quality?: AudioQuality;
}

File: backend/app/schemas/stats.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel

from .album import AlbumSummary
from .pagination import PaginatedResponse
from .track import TrackWithRelations


class LibraryOverview(BaseModel):
    total_albums: int
    total_tracks: int
    total_artists: int
    total_duration_seconds: int
    total_duration_hours: float
    estimated_size_mb: float
    average_bitrate: float | None = None
    average_track_duration: Optional[float] = None
    recently_added_at: Optional[datetime] = None


class GenreDistribution(BaseModel):
    genre: str
    track_count: int
    total_duration_seconds: float


class TrackPlayStats(TrackWithRelations):
    play_count: int
    last_played_at: Optional[datetime] = None


class TrackPlayStatsResponse(PaginatedResponse[TrackPlayStats]):
    """Paginated response for most played tracks."""


class ArtistStats(BaseModel):
    name: str
    track_count: int
    album_count: int


class AlbumStats(BaseModel):
    title: str
    artist_name: Optional[str] = None
    release_date: Optional[str] = None
    track_count: int
    total_duration_seconds: int
    total_duration_minutes: int


class BitrateDistribution(BaseModel):
    bitrate: int
    count: int


class SampleRateDistribution(BaseModel):
    sample_rate: int
    count: int


class QualityStats(BaseModel):
    bitrate_distribution: list[BitrateDistribution]
    sample_rate_distribution: list[SampleRateDistribution]


class DashboardData(BaseModel):
    overview: LibraryOverview
    recent_albums: list[AlbumSummary]
    top_tracks: list[TrackPlayStats]
    genre_distribution: list[GenreDistribution]
    scan_status: str
    scan_progress: Optional[float] = None
    generated_at: datetime


__all__ = [
    "AlbumStats",
    "ArtistStats",
    "BitrateDistribution",
    "DashboardData",
    "GenreDistribution",
    "LibraryOverview",
    "QualityStats",
    "SampleRateDistribution",
    "TrackPlayStatsResponse",
    "TrackPlayStats",
]
